packages/runtime/src/components/page/HeadSnippet.tsx

```diff
@@ -0,0 +1,122 @@
+import { Children, createElement, useEffect, useRef } from 'react'
+import { Snippet } from '../hooks/usePageSnippets'
+import { useServerInsertedHTML } from 'next/navigation'
+import parse from 'html-react-parser'
+import Head from 'next/head'
+import { useIsPagesRouter } from '../../next/hooks/use-is-pages-router'
+
+type Props = {
+  snippet: Snippet
+}
+
+export function HeadSnippet({ snippet }: Props) {
+  const isPagesRouter = useIsPagesRouter()
+
+  useEffect(() => {
+    if (isPagesRouter) {
+      return () => {
+        const snippetElements = getSnippetElementsFromDOM(snippet)
+        if (snippetElements.length > 0) return
+        cleanUpSnippet(snippet)
+      }
+    }
+
+    const snippetElements = getSnippetElementsFromDOM(snippet)
+    const isAlreadyInTheDOM = snippetElements.length > 0
+
+    if (isAlreadyInTheDOM) return
+
+    return renderSnippetAndExecuteScripts(snippet, document.head)
+  }, [isPagesRouter, snippet])
+
+  const headSnippetElement = snippetToElement(snippet)
+  const insertedServerHTML = useRef(false)
+
+  useServerInsertedHTML(() => {
+    if (isPagesRouter || insertedServerHTML.current) return
+
+    insertedServerHTML.current = true
+
+    return headSnippetElement
+  })
+
+  if (isPagesRouter) {
+    return <Head>{headSnippetElement}</Head>
+  }
+
+  return null
+}
+
+const SNIPPET_ID_ATRIBUTE_NAME = 'data-makeswift-snippet-id'
+
+const VALID_HEAD_ELEMENT_TYPES = [
+  'title',
+  'base',
+  'link',
+  'style',
+  'meta',
+  'script',
+  'noscript',
+  'template',
+]
+
+function snippetToElement(snippet: Pick<Snippet, 'id' | 'code'>): (string | JSX.Element)[] {
+  return Children.map(parse(snippet.code), element => {
+    if (typeof element === 'string') return element
+
+    if (!VALID_HEAD_ELEMENT_TYPES.includes(element.type as string)) return null
+
+    const key = element.key ? `${snippet.id}:${element.key}` : snippet.id
+
+    return createElement(element.type, {
+      ...element.props,
+      key,
```

Comment:
Without this key, `next/head` will only insert one snippet into the head. If there are multiple snippets, only the first will get rendered. It's a very strange behavior.

Reply:
Makes sense. But then we still have the issue of duplicate keys. When is `element.key` not defined? Does it always fall back to `snippet.id`?

Reply:
Yes, it will always fall back on `snippet.id`. In fact, this key is what `next/head` uses to dedupe script executions, so it is something we want to be one-to-one with a snippet.

Reply:
The potential issue, though, is a snippet with multiple elements using the same key for all elements